#+TITLE: Sequence Alignment
#+OPTIONS: toc:nil

* Dotplots

Dotter is a graphical dotplot program for detailed comparison of two sequences.
It generates a dotplot where the first sequence to be compared is set along the
x-axis, and the second sequence along the y-axis. If two sequences are similar
in two regions, a dot is plotted with an intensity proportional to the
similarity.

** dotplot programs:
- Desktop applications:
  - dotter (https://www.sanger.ac.uk/science/tools/seqtools), linux only
  - gepard (http://cube.univie.ac.at/gepard)  - suitable for larger sequences, platform independent
  - re-DOT-able https://www.bioinformatics.babraham.ac.uk/projects/redotable/
- Web applications
  - YASS server (https://bioinfo.lifl.fr/yass/index.php), suitable for large sequences or whole chromosomes
  - D-genies (http://dgenies.toulouse.inra.fr/) - whole genome comparison
  
    
*** Dotter
To run *dotter*, open terminal window and  type "dotter" followed by two file names:
#+BEGIN_SRC bash
dotter file1.fasta file2.fasta
#+END_SRC
- Sequences from dotter program must be provided in [[./file_formats.org#the-fasta-format][FASTA format]]
- Dotter can be used for protein-proten, DNA-DNA or protein-DNA comparison

*** Dotter keyboard shortcuts
to navigate in =dotter= program use mouse click to select region
- To move cursor in dotter use *arrows* for vertical or diagonal direction
- to move diagonally use ~<~, ~>~, ~[~ or  =]=. 
- To zoom select area with mouse while holding middle mouse button.

*** Gepard 
Gepard is Java based program, also available on Windows or MacOS. See https://cube.univie.ac.at/gepard
To run gepard:
#+begin_src 
conda activate gepard
gepard
#+end_src
In Graphical user interface select pair of sequences you want to compare. Test different word size settings

** Exercise 1 - structure of HER proteins - Human epidermal growth receptors

- download protein sequences of receptor protein-tyrosine kinase from Uniprot:
| accessions | name        |
|------------+-------------|
| P00533     | ERBB2       |
| P21860     | ERBB3       |
| Q15303     | ERBB4       |
| O18735     | ERBB2 (Dog) |
|------------+-------------|

data for this exercise could  be downloaded from directly from uniprot using the wget command:
#+begin_src sh 
mkdir -p ~/data/dotter_sequences
cd ~/data/dotter_sequences
wget https://www.uniprot.org/uniprot/P00533.fasta -O ERBB2.fasta
wget https://www.uniprot.org/uniprot/P21860.fasta -O ERBB3.fasta
wget https://www.uniprot.org/uniprot/Q15303.fasta -O ERBB4.fasta
wget https://www.uniprot.org/uniprot/O18735.fasta -O ERBB2_dog.fasta
# check sequence statistics:
seqkit stat *.fasta
#+end_src

- compare /ErbB2/ against /ErB3/. Notice the patterns
  in the dotplot and try to find functional domains, for example cysteine rich
  regions B
- try different dotplot stringency using slider on gray-scale strip
- sequences below dotplot correspond to position of blue cross, you can change
  the position of cross either using mouse or  by keyboard shortcuts.
- click on the line and then use arrows to find a good alignment.
- when you identify match, move along diagonal.
- compare /ErbB2/ with /ErbB2-dog/. Do you see the same pattern? 
- compare all proteins to all proteins (concatenate all four FASTA file into one sequence using ~cat~ command)

  
  #+begin_comment
it is necessary to use ubuntu interface!
explain double click and copying
make aha slide for this task
  #+end_comment



#+begin_src sh 
dotter ERBB2.fasta ERBB3.fasta # human ERBB2 vs human ERB3  (paralogs)
dotter ERBB2.fasta ERBB2_dog.fasta  # human ERBB2 vs dog ERBB2  (orthologs)
# all to all comparison:
cat E*.fasta > all_erb.fasta # first we need concatenated sequences in single fasta file
dotter all_erb.fasta all_erb.fasta
#+end_src

Domain structure of *ERBB2* protein:
[[../fig/ERBB2.png]]


Domain structure of *ERBB3* protein:
[[../fig/ERBB3.png]]



#+BEGIN_COMMENT
backup of sequences is stored in ~/Desktop/bioinformatics/data/dotter_sequences/
insertion - KTLSP
#+END_COMMENT

** Exercise 2 - comparison of HOX proteins
Download sequence for protein from uniprote, concatenate all HOX protein
sequence into single FASTA file and make all-to-all comparison using dotplot.

#+begin_src 
wget https://rest.uniprot.org/uniprotkb/P49639.fasta -O HOXA1.fasta
wget https://rest.uniprot.org/uniprotkb/P20719.fasta -O HOXA5.fasta
wget https://rest.uniprot.org/uniprotkb/P09067.fasta -O HOXB5.fasta
wget https://rest.uniprot.org/uniprotkb/P14653.fasta -O HOXB1.fasta

cat H*.fasta > all_hox.fasta
dotter all_hox.fasta all_hox.fasta

#+end_src
Which paires of protein are more similar to each other?

** Exercise 3 - Locate exon/intron boundaries using dotter (splice sites).
- download sequence AC108130.3 from genbank, save only region from 60000 to 119999 
- download cDNA sequence of GABA A receptor: https://www.ncbi.nlm.nih.gov/nuccore/21265167?report=fasta
- run dotter on these two sequences, identify exon/intron structure
- Are the splice sites consensus splice sites? In vertebrates, the intron starts
  with GT and ends with AG, which are called consensus splice sites.
- What's going on at the 3' end of the cDNA?
- download GABAA1 protein sequence - https://www.ncbi.nlm.nih.gov/protein/27808653?report=fasta
- make dotter of genomic dna vs protein sequence
- what is different, why is the protein alignment shorter than cDNA

#+begin_src sh
dotter AC108130.3.fna BC030696.1.fna  # genome vs cDNA
dotter AC108130.3.fna GBRA1_HUMAN.fna   # genome vs protein 
#+end_src
# Data are also available in ~/Desktop/bioinformatics/data/dotter_sequences
When using =dotter= to compare  DNA to protein, DNA sequence must be in forward orientation!

** Exercise 4 - Identification of repeats

*** example repeats in DNA sequences 
**** Inverted repeat
 Triticum aestivum DNA, mobile element MITE contains inverted repeat. Compare the
 sequences of MITE element against itself. 
- What is the position of inverted repeat?
- can you identify palindromes in the sequence? 
#+begin_src bash
# copy sequence to current directory
cp ~/Desktop/Bioinformatics/data/dotter_sequences/inverted_repeat.fasta .
# make dotplot
dotter inverted_repeat.fasta inverted_repeat.fasta
#+end_src

**** Direct repeat
Boundary of transposable element is defined by long terminal repeat (LTR).
Make dotplot of sequence which contain LTR retrotransposon against itself. 
- What is the length of the whole LTR retrotransposon?
- What is the length of LTR?
#+begin_src bash
cp ~/Desktop/Bioinformatics/data/dotter_sequences/direct_repeat.fasta .
dotter direct_repeat.fasta direct_repeat.fasta
#+end_src
**** Tandem repeat
Make dotplot of sequence containing tandem repeat.
- how do you interpret dotplot?
- what is a monomer length of tandem repeat
Hint: distances between diagonal parallel lines can be used to estimate length of monomer in tandem repeat.
#+begin_src bash
cp ~/Desktop/bioinformatics/data/dotter_sequences/tandem_repeat.fasta .
dotter tandem_repeat.fasta tandem_repeat.fasta
#+end_src

** Exercise 5 - Identification of insertions, deletions, duplications
Compare two genomic regions =a_region= and =b_region=
- first do self comparison for each sequence
- then compare =a_region= against =b_region=
- What you can say about these genomic regions? Are there any insertions,
  duplications or deletions?

#+begin_src bash
cp ~/Desktop/bioinformatics/data/dotter_sequences/*_region.fasta .
dotter a_region.fasta a_region.fasta
dotter b_region.fasta b_region.fasta
dotter a_region.fasta b_region.fasta
#+end_src
** Exercise 6 - Overlapping sequences
make dotplot from following sequences stored in file:
#+begin_example
~/Desktop/bioinformatics/data/dotter_sequences/dna_examples/ovoverlaping_sequences.fasta
#+end_example

what is relation between sequences?

solution:
[[../fig/scheme_dotter_overlap.png]]

** Exercise 7 - Compare two genomic regions:
#+begin_src 
cp ~/Desktop//bioinformatics/data/dotter_sequences/genome?_part.fasta ,
dotter genomeA.fasta genomeB.fasta
#+end_src
How would characterize difference between genomes?
** Exercise - Identification of problems in sequences from SRA database
*** Illumina data : SRR2911427  (Migratory locust WGS)
# overlap, high NN proportion
- download sequences from SRA database using =fastq-dump= command line program 
- for documentation see https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump )
- inspect data and create dotplot

#+BEGIN_SRC bash
mkdir ~/tmp
cd ~/tmp
fastq-dump -X 20 --split-files --fasta SRR2911427
# -X 20 = download 20 sequences only
# --fasta = convert sequences to fasta format
# --split-file = create two files one for each pair
# SRR291142 - accession ID
ls -l
cat SRR2911427_1.fasta
dotter SRR2911427_1.fasta SRR2911427_2.fasta
#+END_SRC

what does it mean? 
# paired sequences are overlaping

*** Illumina data : SRR453021 (Nicotian repanda - WGS)
# adapter sequence
# ACACTCTTTCCCTACACGACGCTCTTCCGATCT
# change it to faster-dump!
#+BEGIN_SRC bash
fastq-dump -X 50 --split-files  --fasta SRR453021
dotter SRR453021_2.fasta SRR453021_2.fasta 
dotter SRR453021_1.fasta SRR453021_1.fasta 
dotter SRR453021_1.fasta SRR453021_2.fasta 
#+END_SRC
select repeated sequences using dotter and search with NCBI blast:

https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome

#+begin_comment
# sequence 
** Whole genome/chromosome dotplot

- download part of genome assembly from Zea mays:
- link   https://www.ncbi.nlm.nih.gov/nuccore/KV887992.1
- use range 1-1,000,000 nt
- start GUI program =gepard=
- as sequences 1 and 2 select downloaded sequence and create dotplot
- what do you see and what does it mean?

#+end_comment

* Pairwise alignment
** Sequences for alignment:
*** =ERB2_HUMAN= : http://www.uniprot.org/uniprot/P04626.fasta   
*** =EGFR_DROME= : http://www.uniprot.org/uniprot/P04412.fasta   
*** =Unknown protein= : http://www.uniprot.org/uniprot/Q8SZW1.fasta
(Receptor tyrosine-protein kinase erbB-2, Epidermal growth factor receptor)

To download sequences use either web browser of try to use ~wget~ command in terminal:
#+BEGIN_SRC bash
cd ~/Downloads
wget http://www.uniprot.org/uniprot/P04626.fasta
wget http://www.uniprot.org/uniprot/P04412.fasta
wget http://www.uniprot.org/uniprot/Q8SZW1.fasta
#+END_SRC

** Exercise - compare global and local alignments
- global alignment is performed by program ~needle~
  - http://www.bioinformatics.nl/cgi-bin/emboss/help/needle
- for local alignment use program ~water~, 
  - http://www.bioinformatics.nl/cgi-bin/emboss/help/water

- Programs ~needle~ and ~water~ are available from command line or from EBI web interface: http://www.ebi.ac.uk/Tools/emboss/
- Sequences for alignments are located in directory =~/Desktop/bioinformatics/data/alignment_sequences=
- compare ERB2 (P04626.fasta) vs EGFR (P04412.fasta) using ~needle~ and then using ~water~ using command lne programs:
#+BEGIN_SRC bash
# command example:
needle P04626.fasta P04412.fasta
water P04626.fasta P04412.fasta
#+END_SRC
same programs are also available from web interface:
- https://www.ebi.ac.uk/Tools/psa/emboss_water/
- https://www.ebi.ac.uk/Tools/psa/emboss_needle/

#+begin_comment
show/ explain alignment options - it is possible to select scoring matrix and gap penalties
#+end_comment

- compare ERB2 (P04626.fasta) vs Unknown protein (Q8SZW1.fasta) using ~needle~ and then using ~water~
- what is difference between local and global alignments?
- what happened what gap penalty is increased to 20 and extend_penalty to 5 when using local alignment
- what happened with global alignment if you change =end gap panalty= setting.
- by default BLOSUM62 scoring matrix is used, what happend when you use PAM10?
- compare these protein sequence using =dotter=
#+begin_comment
using PAM10 - does don really tolerate mismatches, gap penalty would have to be set accordingly 
#+end_comment

#+BEGIN_SRC bash
# command line example using PAM10
water P04626.fasta P04412.fasta -datafile EPAM10
#+END_SRC
*** differences between PAM10 and BLOSUM62 matrices
PAM10 : ftp://ftp.ncbi.nih.gov/blast/matrices/PAM10
BLOSUM62 : ftp://ftp.ncbi.nih.gov/blast/matrices/BLOSUM62

*** using blast (blast2seq) to create local aligment for two sequences:
https://blast.ncbi.nlm.nih.gov/Blast.cgi?BLAST_SPEC=blast2seq&LINK_LOC=align2seq&PAGE_TYPE=BlastSearch
blast2seq can be used instead of =needle=. It also provide graphical view of alignment and non-interactive dotplot. Use blast2 seq on  =P04626.fasta= and  =P04412.fasta= sequences and explore results. Compare alignments and dotplot.

You can paste either AA sequences to the blast form or you can use just accession ID (P046256, P04412).

* Exercise - Multiple sequence alignment - Cyclin-dependent kinase
Cyclin-dependent kinases (CDKs) are a group of enzymes that regulate the
progression of the cell cycle by adding phosphate groups to other proteins, a
process called phosphorylation. They are activated by binding to regulatory
proteins called cyclins, which undergo cyclic changes in abundance and activity
throughout the cell cycle.

Create multiple sequence alignment for group of CDKs from human and mouse. Use
program ~mafft~.  use default setting. Before running
~mafft~ check help documentation using ~mafft --help~


#+begin_src bash
mkdir cdk
cd cdk
cp ~Desktop/bioinformatics/data/alignment_sequences/CDK/cdk.fasta .
mafft --help
mafft cdk.fasta > cdk_fasta.aln

#+end_src
Inspect alignment using Jalview program. 
- Try different coloring schemes-  clustal, percentage identity, hydrophobicity
  and by conservation
- Try to identify the most conserved regions.
- What are the coordinates of most conserved region related to *CDK1* sequence.
- Compare this conserved regions with conserved regions which can be identified
  using *conserved domain database*.  use this ling for search
  https://www.ncbi.nlm.nih.gov/Structure/cdd/wrpsb.cgi and CDK1 sequence below.
#+begin_src text
>CDK1
MEDYTKIEKIGEGTYGVVYKGRHKTTGQVVAMKKIRLESEEEGVPSTAIREISLLKE
LRHPNIVSLQDVLMQDSRLYLIFEFLSMDLKKYLDSIPPGQYMDSSLVKSYLYQILQ
GIVFCHSRRVLHRDLKPQNLLIDDKGTIKLADFGLARAFGIPIRVYTHEVVTLWYRS
PEVLLGSARYSTPVDIWSIGTIFAELATKKPLFHGDSEIDQLFRIFRALGTPNNEVW
PEVESLQDYKNTFPKWKPGSLASHVKNLDENGLDLLSKMLIYDPAKRISGKMALNHP
YFNDLDNQIKKM
#+end_src

* Exercise -  Multiple alignment from HSPB8 proteins
Create MSA for set of orthologs of HSPB8 protein (Heat shock protein beta-8) and identify conserved regions.

Make copy of of fasta file and then rename fasta headers:
#+begin_src bash
mkdir hspb8
cd hsbb8
cp ~/Desktop/bioinformatics/data/alignment_sequences/HSP8.fasta .
gedit HSP8.fasta
#+end_src
Change the header from ~>XP_004934466.1 heat shock protein beta-8 [Gallus gallus]~ to ~>Gallus_gallus~
Rename all sequences in the same way.

Create alignment using ~mafft~ program.
#+begin_src 
mafft HSP8.fasta > HSP8_aln.fasta
#+end_src

Open resulting alignment in ~Jalview~ program.
- Inspect alignment, Try different coloring schemes. (see https://www.jalview.org/help/html/colourSchemes/index.html)
- What part of proteins is conserved?
- compare conserved part with domains annotation
  - go to https://www.ncbi.nlm.nih.gov/protein/NP_055180.1
  - select =analyze this sequence/identify conserved domains=
  - will you be able to identify conserved domain if you use only mouse, cow, pig and human sequences?
  - try to  select subset of sequences(mammals) and create alignment again. 
- save alignment in =fasta= and and =msf= format, inspect the saved file using =less= command
* Exercise - Multiple sequence alignment of the mitochondrial 16S gene
The mitochondrial 16S gene is a widely studied genetic marker in molecular
biology, which is used for species identification, phylogenetic analysis, and
evolutionary studies. 16S gene codes for a RNA subunit of the mitochondrial
ribosome and contains many regions with high substitution rates.
We will use *MAFFT* to align the two sets of sequences, and visualize the
resulting alignments with program called *AliView*. Alignments can be edited
manually or automatically with the software *BMGE*, which determines the most
reliable alignment positions based on the proportion of missing data and their
entropy score.

make new directory and copy sequences. Each sequence is iin one file. We will
concatenate to single multi FASTA file using ~cat~ command

#+begin_src bash
mkdir MSA_16s
cd MSA_16s
cat ~/Desktop/bioinformatics/data/alignment_sequences/16s/*.fasta > 16s.fasta
# get information about sequences
seqkit stat 16s.fasta
#+end_src
Inspect resulting file with ~less~ command. 

Align sequence using ~mafft~ program, at first use default setting. Before running
~mafft~ check help documentation using ~mafft --help~

#+begin_src bash
mafft --help
# align 16s with defaults
mafft 16s.fasta > 16s_aln.fasta
# explore output with less command
less 16s_aln.fasta
seqkit stats 16s_aln.fasta
#+end_src

Inspect alignment using ~Aliview~ program.
#+begin_src bash
~/Desktop/bioinformatics/bin/aliview 16s_aln.fasta
#+end_src
Inspect alignment. By default, ~mafft~ keep order of sequences in the alignment
same as in input file. Close ~aliview~ and rerun ~mafft~ with ~reorder~ option. 
#+begin_src 
mafft --reorder 16s.fasta > 16s_aln.fasta
#+end_src
inspect alignment in ~Aliview~

** Manual editing of alignment : 

In the AliView window of 16s_aln.fasta, place the cursor on the sequence that
bridges the first of the two large gaps near the end of the alignment (around
bp 2000) and zoom in (ctrl + mouse wheel) until you can see the labels of the
individual bases. You’ll see that the taxon responsible for these gaps is called
/Balistecaprisc/. It appears that the sequence alignment for this taxon is correct
up to this gap , but that the sequence is not homologous to other taxa between
bp ~ 2000 and the end of the alignment.

Use the cursor to select position begining of gap around 2000 bp of the
‘Balistecaprisc’ sequence. Use ‘Expand selection Right’ in the ‘Selection’ menu.

Remove this part of the ‘Balistecaprisc’ sequence using ‘Clear selected bases’
in the ‘Edit’ menu or just press delete

After removing this part of the ‘Balistecaprisc’ sequence, the two large gaps
near the end are not bridged by any sequence anymore. Remove these gaps entirely
using ‘Delete gap-only columns’ in the ‘Edit’ menu.

Have a look at the regions which appears to be poorly aligned. Use the cursor to
click in the ruler area (above the alignment) and select the regions delimited
by boundary sites which appear to be reliably aligned, in contrast to the
alignment block between these boundaries.

In the ‘Align’ menu, click ‘Change default Aligner program > for realigning all
(or selected blocks)’.

Click the third radio button to select ‘Mafft-globalpair’ as the default
algorithm for realignment. Make sure that the specified MAFFT installation path
is correct, and confirm with ‘OK’.

Click ‘Realign selected block’ in the ‘Align’ menu.

Does the alignment look more reliable now? Once more, remove gap-only columns,
and save the alignment file.

check the lengths of resulting alignments.

** Automatic evaluation of alignment
*** using ~t_coffee~ program 
(this can take several hours to finish)

TCS is an alignment evaluation score that makes it possible to identify the most
correct positions in an MSA. It has been shown that these positions are the most
likely to be structuraly correct and also the most informative when estimating
phylogenetic trees. The TCS evaluation and filtering procedure is implemented in
the T-Coffee package and can be used to evaluate and filter any third party MSA

The TCS is most informative when used to identify low-scoring portions within an
MSA. It is also worth noting that the TCS is not informative when aligning less
than five sequences.

check is program t_coffee is installed, if not run:
#+begin_src bash
sudo apt-get install t_coffee
#+end_src
evaluate alignment:

#+begin_src bash
# this can take several minutes to finish
t_coffee -infile  16s_aln.fasta -evaluate
#+end_src

*** using ~bmge~ program
BMGE is able to perform biologically relevant trimming on a multiple alignment
of DNA, codon or amino acid sequences. BMGE is designed to select regions in a
multiple sequence alignment that are suited for phylogenetic inference. For each
character, BMGE computes a score closely related to an entropy value.
Calculation of these entropy-like scores is weighted with BLOSUM or PAM
similarity matrices in order to distinguish among biologically expected and
unexpected variability for each aligned character

#+begin_src 
conda create -n bmge -c conda-forge -c biconda bmge
conda activate bmge
#+end_src

run ~bmgi~
#+begin_src 
bmge  -i 16s_aln.fasta -t DNA -of 16s_filtered.fasta -oh 16s_filtered.html
#+end_src
~bmge~ generates two output - html is suitable for viewing in web browser.
fitered.fasta can be used for phylogenetic analysis.


* Amino Acid codes

[[../fig/aa_codes.png]]
* Jalview

- Jalview has *two navigation and editing modes*: _normal mode_, where editing and navigation is
  performed using the mouse, and _cursor mode_ where editing and navigation are performed using
  the keyboard. The *F2* key is used to switch between these two modes.
- Navigation in cursor mode:
  - Jump to Sequence n: Type a number n then press [S] to move to sequence (row) n.
  - Jump to Column n: Type a number n then press [C] to move to column n in the alignment.
  - Jump to Residue n: Type a number n then press [P] to move to residue number
    n in the current sequence.
- Overview of the whole alignment, especially when it is large. Select *View*
⇒ *Overview Window*
- Find sequence - *ctrl-F*



- *Clustal X* color scheme
https://www.jalview.org/help/html/colourSchemes/clustal.html 
- *Blosum62* Gaps are coloured white. If a residue matches the consensus sequence residue at that
position it is coloured dark blue. If it does not match the consensus residue
but the 2 residues have a positive Blosum62 score, it is coloured light blue.
