#+TITLE: Sequence Alignment
#+OPTIONS: toc:nil

* Dotplots

Dotter is a graphical dotplot program for detailed comparison of two sequences.
It generates a dotplot where the first sequence to be compared is set along the
x-axis, and the second sequence along the y-axis. If two sequences are similar
in two regions, a dot is plotted with an intensity proportional to the
similarity.

** dotplot programs:
- Desktop applications:
  - dotter (https://www.sanger.ac.uk/science/tools/seqtools), linux only
  - gepard (http://cube.univie.ac.at/gepard)  - suitable for larger sequences, platform independent
  - re-DOT-able https://www.bioinformatics.babraham.ac.uk/projects/redotable/
- Web applications
  - YASS server (https://bioinfo.lifl.fr/yass/index.php), suitable for large sequences or whole chromosomes
  - D-genies (http://dgenies.toulouse.inra.fr/) - whole genome comparison
  
    
** Dotter
To run *dotter*, open terminal window and  type "dotter" followed by two file names:
#+BEGIN_SRC bash
dotter file1.fasta file2.fasta
#+END_SRC
- Sequences from dotter program must be provided in [[./file_formats.org#the-fasta-format][FASTA format]]
- Dotter can be used for protein-proten, DNA-DNA or protein-DNA comparison

*** Dotter keyboard shortcuts
to navigate in =dotter= program use mouse click to select region
- To move cursor in dotter use *arrows* for vertical or diagonal direction
- to move diagonally use ~<~, ~>~, ~[~ or  =]=. 
- To zoom select area with mouse while holding middle mouse button.


** Exercise 1 - structure of HER proteins - Human epidermal growth receptors

- download protein sequences of receptor protein-tyrosine kinase from Uniprot:
| accessions | name        |
|------------+-------------|
| P00533     | ERBB2       |
| P21860     | ERB3        |
| Q15303     | ERB4        |
| O18735     | ERBB2 (Dog) |
|------------+-------------|

data for this exercise could  be downloaded from directly from uniprot using the wget command:
#+begin_src sh 
mkdir -p data/dotter_sequences
cd data/dotter_sequences
wget https://www.uniprot.org/uniprot/P00533.fasta -O ERBB2.fasta
wget https://www.uniprot.org/uniprot/P21860.fasta -O ERBB3.fasta
wget https://www.uniprot.org/uniprot/Q15303.fasta -O ERBB4.fasta
wget https://www.uniprot.org/uniprot/O18735.fasta -O ERBB2_dog.fasta
# check sequence statistics:
seqkit stat *.fasta
#+end_src

- compare /ErbB2/ against /ErB3/. Notice the patterns
  in the dotplot and try to find functional domains, for example cysteine rich
  regions 
- try different dotplot stringency using slider on gray-scale strip
- sequences below dotplot correspond to position of blue cross, you can change
  the position of cross either using mouse or  by keyboard shortcuts.
- click on the line and then use arrows to find a good alignment.
- when you identify match, move along diagonal.
- compare /ErbB2/ with /ErbB2-dog/. Do you see the same pattern? 
- compare all proteins to all proteins (concatenate all four FASTA file into one sequence using ~cat~ command)

  
  #+begin_comment
it is necessary to use ubuntu interface!
explain double click and copying
make aha slide for this task
  #+end_comment



#+begin_src sh 
dotter ERBB2.fasta ERBB3.fasta # human ERBB2 vs human ERB3  (paralogs)
dotter ERBB2.fasta ERBB2_dog.fasta  # human ERBB2 vs dog ERBB2  (orthologs)
# all to all comparison:
cat *.fasta > all_erb.fasta # first we need concatenated sequences in single fasta file
dotter all_erb.fasta all_erb.fasta
#+end_src

Domain structure of ERBB2 protein
[[../fig/ERBB2.png]]



#+BEGIN_COMMENT
backup of sequences is stored in ~/Desktop/bioinformatics/data/dotter_sequences/
insertion - KTLSP
#+END_COMMENT

** Exercise 2 - comparison of HOX proteins
#+begin_src 
wget https://rest.uniprot.org/uniprotkb/P49639.fasta
wget https://rest.uniprot.org/uniprotkb/P20719.fasta
wget https://rest.uniprot.org/uniprotkb/P09067.fasta

cat P*.fasta > all.fasta
#+end_src

** Exercise 3 - Locate exon/intron boundaries using dotter (splice sites).
- download sequence AC108130.3 from genbank, save only region from 60000 to 119999 
- download cDNA sequence of GABA A receptor: https://www.ncbi.nlm.nih.gov/nuccore/21265167?report=fasta
- run dotter on these two sequences, identify exon/intron structure
- Are the splice sites consensus splice sites? In vertebrates, the intron starts
  with GT and ends with AG, which are called consensus splice sites.
- What's going on at the 3' end of the cDNA?
- download GABAA1 protein sequence - https://www.ncbi.nlm.nih.gov/protein/27808653?report=fasta
- make dotter of genomic dna vs protein sequence
- what is different, why is the protein alignment shorter than cDNA

#+begin_src sh
dotter AC108130.3.fna BC030696.1.fna  # genome vs cDNA
dotter AC108130.3.fna GBRA1_HUMAN.fna   # genome vs protein 
#+end_src
# Data are also available in ~/Desktop/bioinformatics/data/dotter_sequences
When using =dotter= to compare  DNA to protein, DNA sequence must be in forward orientation!

** Exercise - identification of repeats
*** repeat structure in CDC10 protein - ankyrin repeat
- download CDC10 protein - http://www.uniprot.org/uniprot/P01129.fasta 
- make dotplot of CDC10 against itself
- what is the position of repeat?
- compare with position of ANK repeat in uniprot database (https://www.uniprot.org/uniprot/P01129)
#+begin_src sh
wget http://www.uniprot.org/uniprot/P01129.fasta 
dotter P01129.fasta P01129.fasta
#+end_src


#+begin_comment
visibility is quite bad
Repeat1	356 – 385	ANK 1
Repeat2	483 – 512	ANK 2
#+end_comment

*** repeats in ice nucleation protein 
- P20469 is an ice nucleation protein, with a large number of repeats. What is the structure of repeating units
- download sequences from uniprot : http://www.uniprot.org/uniprot/P20469.fasta
#+begin_src sh
wget http://www.uniprot.org/uniprot/P20469.fasta
dotter P20469.fasta P20469.fasta
#+end_src
Distances between diagonal parallel lines can be used to estimate length of monomer in tandem repeat.

*** example repeats in DNA sequences 
**** Inverted repeat
https://www.ncbi.nlm.nih.gov/nuccore/LC052278.1?report=fasta&log$=seqview&format=text
**** Direct repeat
https://www.ncbi.nlm.nih.gov/nuccore/AF049230.1?report=fasta&log$=seqview&format=text
**** Tandem repeat
https://www.ncbi.nlm.nih.gov/nuccore/KF293390.1?report=fasta&log$=seqview&format=text


#+begin_src sh
# sequences are downloaded in ~/Desktop/bioinformatics/data/dotter_sequences/dna_examples
# go to directory with sequences
# run
dotter direct_repeat.fasta direct_repeat.fasta 
dotter direct_repeat.fasta direct_repeat.fasta
dotter inverted_repeat.fasta inverted_repeat.fasta
#+end_src

** overlaping sequences
make dotplot from following sequences stored in file:
#+begin_example
~/Desktop/bioinformatics/data/dotter_sequences/dna_examples/ovoverlaping_sequences.fasta
#+end_example

what is relation between sequences?
 
solution:
[[../fig/scheme_dotter_overlap.png]]

** Exercise - identification of problems in sequences from SRA database
*** Illumina data : SRR2911427  (Migratory locust WGS)
# overlap, high NN proportion
- download sequences from SRA database using =fastq-dump= command line program 
- for documentation see https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump )
- inspect data and create dotplot

#+BEGIN_SRC bash
mkdir ~/tmp
cd ~/tmp
fastq-dump -X 20 --split-files --fasta SRR2911427
# -X 20 = download 20 sequences only
# --fasta = convert sequences to fasta format
# --split-file = create two files one for each pair
# SRR291142 - accession ID
ls -l
cat SRR2911427_1.fasta
dotter SRR2911427_1.fasta SRR2911427_2.fasta
#+END_SRC

what does it mean? 
# paired sequences are overlaping

*** Illumina data : SRR453021 (Nicotian repanda - WGS)
# adapter sequence
# ACACTCTTTCCCTACACGACGCTCTTCCGATCT
# change it to faster-dump!
#+BEGIN_SRC bash
fastq-dump -X 50 --split-files  --fasta SRR453021
dotter SRR453021_2.fasta SRR453021_2.fasta 
dotter SRR453021_1.fasta SRR453021_1.fasta 
dotter SRR453021_1.fasta SRR453021_2.fasta 
#+END_SRC
select repeated sequences using dotter and search with NCBI blast:

https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome

#+begin_comment
# sequence 
** whole genome/chromosome dotplot

- download part of genome assembly from Zea mays:
- link   https://www.ncbi.nlm.nih.gov/nuccore/KV887992.1
- use range 1-1,000,000 nt
- start GUI program =gepard=
- as sequences 1 and 2 select downloaded sequence and create dotplot
- what do you see and what does it mean?

#+end_comment

* Pairwise alignment
** Sequences for alignment:
*** =ERB2_HUMAN= : http://www.uniprot.org/uniprot/P04626.fasta   
*** =EGFR_DROME= : http://www.uniprot.org/uniprot/P04412.fasta   
*** =Unknown protein= : http://www.uniprot.org/uniprot/Q8SZW1.fasta
(Receptor tyrosine-protein kinase erbB-2, Epidermal growth factor receptor)

To download sequences use either web browser of try to use ~wget~ command in terminal:
#+BEGIN_SRC bash
cd ~/Downloads
wget http://www.uniprot.org/uniprot/P04626.fasta
wget http://www.uniprot.org/uniprot/P04412.fasta
wget http://www.uniprot.org/uniprot/Q8SZW1.fasta
#+END_SRC

** Exercise - compare global and local alignments
- global alignment is performed by program ~needle~
  - http://www.bioinformatics.nl/cgi-bin/emboss/help/needle
- for local alignment use program ~water~, 
  - http://www.bioinformatics.nl/cgi-bin/emboss/help/water

- Programs ~needle~ and ~water~ are available from command line or from EBI web interface: http://www.ebi.ac.uk/Tools/emboss/
- Sequences for alignments are located in directory =~/Desktop/bioinformatics/data/alignment_sequences=
- compare ERB2 (P04626.fasta) vs EGFR (P04412.fasta) using ~needle~ and then using ~water~ using command lne programs:
#+BEGIN_SRC bash
# command example:
needle P04626.fasta P04412.fasta
water P04626.fasta P04412.fasta
#+END_SRC
same programs are also available from web interface:
- https://www.ebi.ac.uk/Tools/psa/emboss_water/
- https://www.ebi.ac.uk/Tools/psa/emboss_needle/

#+begin_comment
show/ explain alignment options - it is possible to select scoring matrix and gap penalties
#+end_comment

- compare ERB2 (P04626.fasta) vs Unknown protein (Q8SZW1.fasta) using ~needle~ and then using ~water~
- what is difference between local and global alignments?
- what happened what gap penalty is increased to 20 and extend_penalty to 5 when using local alignment
- what happened with global alignment if you change =end gap panalty= setting.
- by default BLOSUM62 scoring matrix is used, what happend when you use PAM10?
- compare these protein sequence using =dotter=
#+begin_comment
using PAM10 - does don really tolerate mismatches, gap penalty would have to be set accordingly 
#+end_comment

#+BEGIN_SRC bash
# command line example using PAM10
water P04626.fasta P04412.fasta -datafile EPAM10
#+END_SRC
*** differences between PAM10 and BLOSUM62 matrices
PAM10 : ftp://ftp.ncbi.nih.gov/blast/matrices/PAM10
BLOSUM62 : ftp://ftp.ncbi.nih.gov/blast/matrices/BLOSUM62

*** using blast (blast2seq) to create local aligment for two sequences:
https://blast.ncbi.nlm.nih.gov/Blast.cgi?BLAST_SPEC=blast2seq&LINK_LOC=align2seq&PAGE_TYPE=BlastSearch
blast2seq can be used instead of =needle=. It also provide graphical view of alignment and non-interactive dotplot. Use blast2 seq on  =P04626.fasta= and  =P04412.fasta= sequences and explore results. Compare alignments and dotplot.

You can paste either AA sequences to the blast form or you can use just accession ID (P046256, P04412).

* Multiple alignment from HSPB8 proteins
purpose of this exercise is to obtain set of orthologs of HSPB8 protein (Heat shock protein beta-8), create multiple sequence alignment and identify conserved regions.

- obtain protein sequence of HSPB8 from NCBI, 
# gene link is: https://www.ncbi.nlm.nih.gov/gene/26353
- protein link is https://www.ncbi.nlm.nih.gov/protein/NP_055180.1
- from right menu select =analyze this sequence/Run BLAST=
- select search against sefseq_protein database, set algorithm of search to DELTA-BLAST (Domain Enhanced Lookup Time Accelerated BLAST)
- Limit blast search to the following species (add species to =Organims= field):
  - Sus scrofa (pig) 
  - Danio rerio (zebrafish) 
  - Bos taurus (cattle)
  - Mus musculus (house mouse)
  - Xenopus laevis (African clawed frog)
  - Drosophila melanogaster (fruit fly)
  - Gallus gallus (chicken) 
# list of accessions:  	XP_004934466 NP_523827 NP_001094427 NP_001079782 NP_109629 NP_001014955 NP_001094427 XP_001929620 NP_055180.1
- select best hit from each species :
- download fasta file for selected proteins. 
# make sure that other sequences are unselected!
- edit fasta file in text editor - remove gaps, put the name of organism at the beginning of the description line e.g.:
#+BEGIN_EXAMPLE
>chicken XP_004934466.1 PREDICTED: heat shock protein beta-8 [Gallus gallus]
MADSQLPFSCHYPGRRSLRDPFREPGLTSRLLDDDFGLSPFPGDLTADWPDWARPRLTPTWPGPLRARAS
AMAPGYSTRFGGYPESRSPAPTSREPWKVCVNVHSFKPEELTVKTKDGYVEVSGKHEEQQVEGGIVSKNF
TKKIQLPYEVDPITVFASLSPEGLLIIEAPQIPPYQQYGEGGCSGEIPLESPEATCA
>fruitfly NP_523827.1 lethal (2) essential for life, isoform A [Drosophila melanogaster]
MSVVPLMFRDWWDELDFPMRTSRLLDQHFGQGLKRDDLMSSVWNSRPTVLRSGYLRPWHTNSLQKQESGS
TLNIDSEKFEVILDVQQFSPSEITVKVADKFVIVEGKHEEKQDEHGYVSRQFSRRYQLPSDVNPDTVTSS
LSSDGLLTIKAPMKALPPPQTERLVQITQTGPSSKEDNAKKVETSTA
...
#+END_EXAMPLE
** Create multiple sequences alignment:
*** Option 1:  Jalview program
- start jalview program from web: http://www.jalview.org/
- open sequences from file menu.
- align sequences =Web service/Alignment/Mafft with defaults=
 

*** Option 2: Use web browser interface:
- go to http://www.ebi.ac.uk/Tools/msa/clustalo/
- paste your sequences and align
- color alignment or go to result viewer to create visualization using Mview
*** Option 4: Alignment using command line program.

   #+begin_src sh
   mafft HSP8.fasta > aligned_sequences_mafft.fasta
   muscle -in HSP8.fasta -out aligned_sequences_muscle.fasta
   #+end_src
*** Option 4: Geneious Prime program
- start Geneious
- create new project folder
- import sequences from file
- create multiple sequence alignment from Align/Assemble/Multiple align. Use program MUSCLE
- reorder sequences in alignments, try to edit alignment gaps
  

** Inspect MSA results
- Inspect alignment, Try different coloring schemes. (see https://www.jalview.org/help/html/colourSchemes/index.html)
- What part of proteins is conserved?
- compare conserved part with domains annotation
  - go to https://www.ncbi.nlm.nih.gov/protein/NP_055180.1
  - select =analyze this sequence/identify conserved domains=
  - will you be able to identify conserved domain if you use only mouse, cow, pig and human sequences?
  - try to  select subset of sequences(mammals) and create alignment again. 
- save alignment in =fasta= and and =msf= format, inspect the saved file using =less= command

* Amino Acid codes

[[../fig/aa_codes.png]]
