# Comparison of phylogenetic trees based on amino acid and nucleotide sequences of ubiquitin gene
  1. Open Jalview program and load the ubiquitin nucleotide sequences from `~/Desktop/Bioinformatics/data/phylogenetic/ubiquitin.fna`
  2. Create a multiple sequence alignment (MSA) using the ubiquitin nucleotide sequences. You can use mafft program from Jalview interface.
  3. Convert DNA alignment to amino acid alignment using the translation tool in Jalview. Use menu `Calculate > Translate as cDNA`
  4. Calculate Neighbor-Joining (NJ) tree for amino acid and nucleotide sequences. Use menu `Calculate > Calculate Tree of PCA...`
  5. Compare the topology of the trees. What do you observe? Are trees fully resolved?
  6. Export nucleotide and protein trees in Newick format. and visualize them using Dendroscope program
  7. Try different tree visualization options in Dendroscope. Test phylogram, cladogram and radial phylogram. Which tree visualization is the most informative? Which visualization better shows polytomies?

     